### PR TITLE
Fix Elasticsearch DevService network for @QuarkusIntegrationTest for docker-image-building

### LINKED ISSUE #43980: Failing devservices-containing @QuarkusIntegrationTest for `container-image.build=true` on MacOS
### Describe the bug

The `@QuarkusIntegrationTest` in the project, whose final artefact is Jib-built docker image, tests this built image.

If there are dev services required for this integration test, those dependency adresses are passed with properties, e.g. `quarkus.elasticsearch.hosts`. The hosts of these addresses are defined as `localhost`, which is correct for jar testing.

Problem occurs for MacOS. In the MacOS, Docker container `localhost` refers only to "this-container-localhost", which causes failure of reaching devservices via `localhost`.

We can observe this behaviour in the logs:
`2024-10-19 23:38:45,069 INFO  [io.qua.tes.com.DefaultDockerContainerLauncher] (main) Executing "docker run --name quarkus-integration-test-SfkPp -i --rm -p 8081:8081 -p 8444:8444 --net=834f31302c23d79e9deb7a0f691fad3f15d419301763337c2901b96d269c2ce2 --env QUARKUS_LOG_CATEGORY__IO_QUARKUS__LEVEL=INFO --env QUARKUS_HTTP_PORT=8081 --env QUARKUS_HTTP_SSL_PORT=8444 --env TEST_URL=http://localhost:8081 --env QUARKUS_ELASTICSEARCH_HOSTS=localhost:50383 code-with-quarkus:1.0.0-SNAPSHOT"`

### Expected behavior

The tested container should access devservices via container's domain name (e.g. `elasticsearch:9200` instead of `localhost:<mappedPort>`) in shared network.

Other workaround specific for MacOS could be replacing `localhost:<mappedPort>` with `host.docker.internal:<mappedPort>`.

### Actual behavior

Accessing devservices on MacOS within `Docker container` via `localhost` causes `java.net.ConnectException: Connection refused`. 

### How to Reproduce?

Reproducing project: [code-with-quarkus.zip](https://github.com/user-attachments/files/17447422/code-with-quarkus.zip)

Steps to create such project:
1. Create example project `quarkus create app`
2. Add docker image building extension - `quarkus extension add container-image-jib`
3. Require building docker image by specifying Maven property `<quarkus.container-image.build>true</quarkus.container-image.build>`
4. Require running IT tests by specifying Maven property `<skipITs>false</skipITs>`
5. Add example required devservice e.g. 
```
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-elasticsearch-rest-client</artifactId>
        </dependency>
```
6. Use devservice in application
7. Run `mvn clean install`

### Output of `uname -a` or `ver`

Darwin MacBook-Pro.local 23.5.0 Darwin Kernel Version 23.5.0: Wed May  1 20:09:52 PDT 2024; root:xnu-10063.121.3~5/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "21" 2023-09-19 OpenJDK Runtime Environment (build 21+35-2513) OpenJDK 64-Bit Server VM (build 21+35-2513, mixed mode, sharing)

### Quarkus version or git rev

3.15.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.4 (dfbb324ad4a7c8fb0bf182e6d91b0ae20e3d2dd9) Java version: 21, vendor: Oracle Corporation Default locale: en_GB, platform encoding: UTF-8 OS name: "mac os x", version: "14.5", arch: "x86_64", family: "mac"

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jgardo**
Fixes #43980 

Tested manually for project attached to https://github.com/quarkusio/quarkus/issues/43980

Unfortunately I couldn't find any test verifying @QuarkusIntegrationTest of created docker image...

Such tests occurred in the past, but they were dropped in PR #31295 . I suppose these tests were dropped intentionally, as time of CI build was too long.

### CODE REVIEW

**Review by yrodiere — APPROVED** [review]
LGTM, thank you.

> Unfortunately I couldn't find any test verifying @QuarkusIntegrationTest of created docker image...

[We do test Elasticsearch dev services](https://github.com/quarkusio/quarkus/blob/95a3a6c1b33e83ae2ee03368a0451cb9497f3b55/extensions/elasticsearch-rest-client/deployment/src/test/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/DevServicesElasticsearchDevModeTestCase.java), but indeed not with `@QuarkusIntegrationTest`.

I would personally be in favor of just using dev services in most integration tests, e.g. in `integration-tests/elasticsearch-rest-client`, instead of relying on [separate configuration](https://github.com/quarkusio/quarkus/blob/c679d63747a2656396d2aaf54e43d6277a5e5324/integration-tests/elasticsearch-rest-client/pom.xml#L168-L235).
But it's certainly outside the scope of this PR, so please don't do that here, and there might be reasons not to do it -- @geoand would know.

File: extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java [hunk 1]
```diff
@@ -209,11 +210,15 @@ private DevServicesResultBuildItem.RunningDevService startElasticsearchDevServic
             container.withReuse(config.reuse);
 
             container.start();
+
+            var httpHost = container.getHost() + ":" + container.getMappedPort(ELASTICSEARCH_PORT);
+            if (createdContainer.hostName() != null) {
+                httpHost = createdContainer.hostName() + ":" + ELASTICSEARCH_PORT;
```
Comment (yrodiere) [inline extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:216]: On second thought... Shouldn't it be this?

```suggestion
                httpHost = createdContainer.hostName() + ":" + container.getMappedPort(ELASTICSEARCH_PORT);
```

**Review by yrodiere — CHANGES_REQUESTED** [review]

**Comment by geoand** [comment]
> But it's certainly outside the scope of this PR, so please don't do that here, and there might be reasons not to do it -- @geoand would know.

Totally agree. Dev services work just fine with QuarkusIntegrationTest

**Comment by yrodiere** [comment]
> > But it's certainly outside the scope of this PR, so please don't do that here, and there might be reasons not to do it -- @geoand would know.
> 
> Totally agree. Dev services work just fine with QuarkusIntegrationTest

There might be a misunderstanding here, because that's not what I meant at all. Anyway, I created #44124 to discuss this.

**Comment by gsmet** [comment]
FWIW, I remember having fixed something similar in another dev service. I wonder if it's something we need to fix globally. Definitely not in the scope of this PR, just thinking out loud.

**Comment by gsmet** [comment]
To be clear, I was thinking about this PR https://github.com/quarkusio/quarkus/pull/42472/files, where we fix a slightly different issue for MongoDB but it ends up adjusting the same values for different reasons.

File: extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java (hunk 1 above)
Comment (jgardo) [reply extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:216]: Container's port within shared network is accessible via direct port without mapping.

Similar code occurs in properly working `extensions/devservices/postgresql/src/main/java/io/quarkus/devservices/postgresql/deployment/PostgresqlDevServicesProcessor.java`. There we can find similar code with fixed port:
```java
        public String getEffectiveJdbcUrl() {
            if (useSharedNetwork) {
                String additionalUrlParams = constructUrlParameters("?", "&");
                return "jdbc:postgresql://" + hostName + ":" + POSTGRESQL_PORT
                        + "/" + getDatabaseName() + additionalUrlParams;
            } else {
                return super.getJdbcUrl();
            }
        }
```

**Comment by jgardo** [comment]
> To be clear, I was thinking about this PR https://github.com/quarkusio/quarkus/pull/42472/files, where we fix a slightly different issue for MongoDB but it ends up adjusting the same values for different reasons.

I suppose there are some devservices to fix (at least these using `io.quarkus.devservices.common.ConfigureUtil#configureSharedNetwork` without storing result hostName).

Personally I also would need to fix Pulsar devservice. So should I create another issue and PR or with same issue with separate PR?

File: extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java (hunk 1 above)
Comment (yrodiere) [reply extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java:216]: Makes sense, thanks.

**Review by yrodiere — APPROVED** [review]

**Pull request merged**